drivers/soc/qcom/memshare.c

```diff
@@ -0,0 +1,483 @@
+// SPDX-License-Identifier: GPL-2.0
+
+// TODO drop NOTE: https://github.com/msm8916-mainline/linux/blob/gps/samples/qmi/qmi_sample_client.c
+
+// FIXME some of those are unused?
+#include <linux/kernel.h>
+#include <linux/types.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/of_address.h>
+#include <linux/mutex.h>
+
+#include <linux/device.h>
+#include <linux/platform_device.h>
+#include <linux/qrtr.h>
+#include <linux/net.h>
+#include <linux/completion.h>
+#include <linux/idr.h>
+#include <linux/string.h>
+#include <net/sock.h>
+#include <linux/soc/qcom/qmi.h>
+
+#include "memshare_qmi_msg.h"
+
+#define MEMSHARE_MAX_CLIENTS 10
+
+struct memshare_client {
+	u32 id;
+	u32 proc_id;
+	u32 qrtr_node;
+	u32 size;
+	phys_addr_t phy_addr;
+};
+
+struct memshare {
+	struct device *dev;
+	struct qmi_handle qmi;
+	unsigned int client_cnt;
+	struct memshare_client *legacy_client;
+	struct memshare_client *clients[MEMSHARE_MAX_CLIENTS];
+};
+
+static int memshare_get_client_index(struct memshare *share, u32 id, u32 proc_id)
+{
+	int i;
+
+	for (i = 0; i < share->client_cnt; i++)
+		if (share->clients[i]->id == id && share->clients[i]->proc_id == proc_id)
+			return i;
+
+	return -EINVAL;
+}
+
+static int memshare_get_client_index_by_address(struct memshare *share, u64 handle)
+{
+	int i;
+
+	if (!handle)
+		return -EINVAL;
+
+	for (i = 0; i < share->client_cnt; i++)
+		if (share->clients[i]->phy_addr == handle)
+			return i;
+
+	return -EINVAL;
+}
+
+static void memshare_alloc_req_handler(struct qmi_handle *qmi, struct sockaddr_qrtr *sq,
+				       struct qmi_txn *txn, const void *data)
+{
+	struct memshare *share = container_of(qmi, struct memshare, qmi);
+	const struct mem_alloc_req_msg_v01 *req = data;
+	struct mem_alloc_resp_msg_v01 resp = { .resp = QMI_RESULT_FAILURE_V01 };
+	struct memshare_client *client;
+	int ret;
+
+	dev_warn(share->dev, "Got alloc_req message:\n"); // TODO drop
+	dev_warn(share->dev, "s=%d balv=%d bal=%d\n",
+		req->num_bytes,
+		req->block_alignment_valid,
+		req->block_alignment
+	);
+
+	if (!share->legacy_client) {
+		dev_err(share->dev, "Unknown request from legacy client (size=%d, node=%d)\n",
+			req->num_bytes, sq->sq_node);
+		goto send_response;
+	}
+
+	client = share->legacy_client;
+
+	if (sq->sq_node != client->qrtr_node) {
+		dev_err(share->dev, "Request from node %d but %d expected\n",
+			sq->sq_node, client->qrtr_node);
+		goto send_response;
+	}
+
+	if (client->size && client->size != req->num_bytes) {
+		dev_err(share->dev, "Request with wrong size was ignored (size=%d)\n",
```

Comment:
```suggestion
		dev_err(share->dev, "Request with wrong size (size=%d)\n",
```

I guess it's not ignored anymore since you send an error response?